core/src/main/java/fr/sncf/osrd/envelope_sim/TrainPhysicsIntegrator.java

```diff
@@ -75,30 +82,89 @@ private IntegrationStep step(double timeStep, double position, double speed) {
         return newtonStep(timeStep, speed, acceleration, directionSign);
     }
 
-    /** Compute the weight force of a rolling stock at a given position on a given path */
-    public static double getWeightForce(PhysicsRollingStock rollingStock, PhysicsPath path, double headPosition) {
+    /** Compute the average grade of a rolling stock at a given position on a given path */
+    public static double getAverageGrade(PhysicsRollingStock rollingStock, PhysicsPath path, double headPosition) {
         var tailPosition = Math.min(Math.max(0, headPosition - rollingStock.getLength()), path.getLength());
         headPosition = Math.min(Math.max(0, headPosition), path.getLength());
-        var averageGrade = path.getAverageGrade(tailPosition, headPosition);
+        return path.getAverageGrade(tailPosition, headPosition);
+    }
+
+    /** Compute the lowest grade of a rolling stock at a given position on a given path,
+     * corresponding to the most downhill part under the train */
+    public static double getLowestGrade(PhysicsRollingStock rollingStock, PhysicsPath path, double headPosition) {
+        var tailPosition = Math.min(Math.max(0, headPosition - rollingStock.getLength()), path.getLength());
+        headPosition = Math.min(Math.max(0, headPosition), path.getLength());
+        return path.getLowestGrade(tailPosition, headPosition);
+    }
+
+    /** Compute a weight force given a rolling stock and a grade value, in m/km */
+    public static double getWeightForce(PhysicsRollingStock rollingStock, double grade) {
         // get an angle from a meter per km elevation difference
         // the curve's radius is taken into account in meanTrainGrade
-        var angle = Math.atan(averageGrade / 1000.0);  // from m/km to m/m
-        return -rollingStock.getMass() * 9.81 * Math.sin(angle);
+        var angle = grade / 1000.0;  // from m/km to m/m
+        return -rollingStock.getMass() * 9.81 * angle;
+    }
+
+    private double getAcceleration(
+            double weightForce,
+            double speed,
+            double tractionForce,
+            double brakingForce
+    ) {
+        // in case of braking, several cases are possible
+        if (action == Action.BRAKE) {
+            if (useCase == TIMETABLE)
+                return rollingStock.getTimetableDeceleration();
+            if (useCase == ETCS_EBD)
+                return (- rollingStock.getSafeBrakingForce(speed) + weightForce) / rollingStock.getInertia();
+            if (useCase == ETCS_SBD)
+                return (- rollingStock.getServiceBrakingForce(speed) + weightForce) / rollingStock.getInertia();
+            if (useCase == ETCS_GUI)
+                return (- rollingStock.getNormalServiceBrakingForce(speed)
+                        + weightForce
+                        + rollingStock.getGradientCorrection(weightForce, speed))
+                        / rollingStock.getInertia();
```

Comment:
I'd prefer a switch (maybe with a default to timetable), otherwise we will forget to change this if we add an enum value